app/drivers/kscan/kscan_gpio_matrix.c

```diff
@@ -1,48 +1,155 @@
+struct kscan_gpio_list {
+    const struct kscan_gpio_dt_spec *gpios;
+    size_t len;
+};
+
+/** Define a kscan_gpio_list from a compile-time GPIO array. */
+#define KSCAN_GPIO_LIST(gpio_array)                                                                \
+    ((struct kscan_gpio_list){.gpios = gpio_array, .len = ARRAY_SIZE(gpio_array)})
+
+struct kscan_matrix_config {
+    struct kscan_gpio_list rows;
+    struct kscan_gpio_list cols;
+    struct kscan_gpio_list inputs;
+    struct kscan_gpio_list outputs;
+    int32_t debounce_period_ms;
+    int32_t poll_period_ms;
+    enum kscan_diode_direction diode_direction;
+};
+
+/**
+ * Get the index into a matrix state array from a row and column.
+ */
+static int state_index_rc(const struct kscan_matrix_config *config, const int row, const int col) {
+    __ASSERT(row < config->rows.len, "Invalid row %i", row);
+    __ASSERT(col < config->cols.len, "Invalid column %i", col);
+
+    return (col * config->rows.len) + row;
+}
+
+/**
+ * Get the index into a matrix state array from input/output pin indices.
+ */
+static int state_index_io(const struct kscan_matrix_config *config, const int input_idx,
+                          const int output_idx) {
+    return (config->diode_direction == KSCAN_ROW2COL)
+               ? state_index_rc(config, output_idx, input_idx)
+               : state_index_rc(config, input_idx, output_idx);
+}
+
+static int kscan_matrix_set_all_outputs(const struct device *dev, const int value) {
+    const struct kscan_matrix_config *config = dev->config;
+
+    for (int i = 0; i < config->outputs.len; i++) {
+        const struct kscan_gpio_dt_spec *gpio = &config->outputs.gpios[i];
+
+        int err = gpio_pin_set(gpio->port, gpio->pin, value);
+        if (err) {
+            LOG_ERR("Failed to set output %i to %i: %i", i, value, err);
+            return err;
+        }
+    }
+
+    return 0;
+}
 
-#if !defined(CONFIG_ZMK_KSCAN_MATRIX_POLLING)
-static int kscan_gpio_config_interrupts(const struct device **devices,
-                                        const struct kscan_gpio_item_config *configs, size_t len,
-                                        gpio_flags_t flags) {
-    for (int i = 0; i < len; i++) {
-        const struct device *dev = devices[i];
-        const struct kscan_gpio_item_config *cfg = &configs[i];
+#if USE_INTERRUPTS
+static int kscan_matrix_interrupt_configure(const struct device *dev, const gpio_flags_t flags) {
+    const struct kscan_matrix_config *config = dev->config;
 
-        int err = gpio_pin_interrupt_configure(dev, cfg->pin, flags);
+    for (int i = 0; i < config->inputs.len; i++) {
+        const struct kscan_gpio_dt_spec *gpio = &config->inputs.gpios[i];
 
+        int err = gpio_pin_interrupt_configure(gpio->port, gpio->pin, flags);
         if (err) {
-            LOG_ERR("Unable to enable matrix GPIO interrupt");
+            LOG_WRN("Unable to configure interrupt for pin %u on %s", gpio->pin, gpio->port->name);
```

Comment:
Any reason this isn't still an `LOG_ERR`?

Reply:
No idea why I changed this. Changed it back.